src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs

```diff
@@ -366,6 +370,35 @@ private void EmitPointerTypeSignature(PointerType type, SignatureContext context
             EmitTypeSignature(type.ParameterType, context);
         }
 
+        private void EmitFunctionPointerTypeSignature(FunctionPointerType type, SignatureContext context)
+        {
+            CorCallingConvention callingConvention = (type.Signature.Flags & MethodSignatureFlags.UnmanagedCallingConventionMask) switch
+            {
+                MethodSignatureFlags.None => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_DEFAULT,
+                MethodSignatureFlags.UnmanagedCallingConventionCdecl => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_C,
+                MethodSignatureFlags.UnmanagedCallingConventionStdCall => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_STDCALL,
+                MethodSignatureFlags.UnmanagedCallingConventionThisCall => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_THISCALL,
+                MethodSignatureFlags.CallingConventionVarargs => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_VARARG,
+                MethodSignatureFlags.UnmanagedCallingConvention => CorCallingConvention.IMAGE_CEE_CS_CALLCONV_UNMANAGED,
+                _ => throw new NotSupportedException()
+            };
+
+            if ((type.Signature.Flags & MethodSignatureFlags.Static) == 0)
+            {
+                callingConvention |= CorCallingConvention.IMAGE_CEE_CS_CALLCONV_HASTHIS | CorCallingConvention.IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS;
```

Comment:
Is `IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS` right here?

Reply:
Hmm, that's exactly what I'm trying to get my head around. I believe that ECMA managed method signatures do include the <code>this</code> pointer so I would assume it's explicit but I may be misunderstanding the purpose of the flag, it's not used very broadly in the CoreCLR native runtime.

Reply:
The MethodSignature initialization only uses `HasThis` to initialize the `Static` bit:

https://github.com/dotnet/runtime/blob/83bd401b112f2deb0db516300f72eb9d1579751d/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs#L399-L400

so the reverse should do the same.

If we need to preserve HasExplicitThis too (I do not think we need to), we would need to store it in MethodSignature.

Reply:
Based on your clarification I have removed the <code>ExplicitThis</code> flag when emitting the non-static function pointer signatures.

Reply:
I am still not sure about the ExplicitThis. I think we should probably store it on MethodSignature when we are initializing it from Ecma metadata, store it here; and also send it to the JIT so that the JIT can do the right thing with it.

Reply:
We have number of IL tests that exercise explicit this (look for "explicit instance" in *.il under tests). I am surprised that none of them fails.

Reply:
Ack, yes, we should handle explicitthis, although fixing all the places that work with that may be beyond the scope of what we should try to fix with this PR. I think it will likely be a fairly large fix to a bunch of places unrelated to the general thrust of the code written here.

Reply:
I have created a .NET 8 tracking issue for this task, https://github.com/dotnet/runtime/issues/73407.